File: src/data/courses/dsa/graphs/grid-graph-patterns/flood-fill/code/java/FloodFill.java
// ============================================================
// Flood Fill
// ============================================================
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        int orig = image[sr][sc];
        if (orig == color) return image;
        dfs(image, sr, sc, orig, color);
        return image;
    }
    void dfs(int[][] img, int i, int j, int orig, int nc) {
        if (i<0||i>=img.length||j<0||j>=img[0].length||img[i][j]!=orig) return;
        img[i][j] = nc;
        dfs(img,i+1,j,orig,nc); dfs(img,i-1,j,orig,nc);
        dfs(img,i,j+1,orig,nc); dfs(img,i,j-1,orig,nc);
    }
}

File: src/data/courses/dsa/graphs/grid-graph-patterns/flood-fill/code/python/flood_fill.py
# ============================================================
# Flood Fill
# ============================================================
class Solution:
    def floodFill(self, image, sr, sc, color):
        orig = image[sr][sc]
        if orig == color: return image
        m, n = len(image), len(image[0])
        def dfs(i, j):
            if i<0 or i>=m or j<0 or j>=n or image[i][j]!=orig: return
            image[i][j] = color
            dfs(i+1,j); dfs(i-1,j); dfs(i,j+1); dfs(i,j-1)
        dfs(sr, sc)
        return image
